cmd/format-fs.go

```diff
@@ -0,0 +1,231 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	errors2 "github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/lock"
+)
+
+// FS format version strings.
+const (
+	formatBackendFS   = "fs"
+	formatFSVersionV1 = "1"
+)
+
+// formatFSV1 - structure holds format version '1'.
+type formatFSV1 struct {
+	formatMetaV1
+	FS struct {
+		Version string `json:"version"`
+	} `json:"fs"`
+}
+
+// Used to detect the version of "fs" format.
+type formatFSVersionDetect struct {
+	FS struct {
+		Version string `json:"version"`
+	} `json:"fs"`
+}
+
+// Returns the latest "fs" format.
+func newFormatFSV1() (format *formatFSV1) {
+	f := &formatFSV1{}
+	f.Version = formatMetaVersionV1
+	f.Format = formatBackendFS
+	f.FS.Version = formatFSVersionV1
+	return f
```

Comment:
I remember now, I had tried to do it and it was looking more complicated (because of the way `formatFSV1` struct is defined), here's how it looks if we do it:
```
// Returns the latest "fs" format.
func newFormatFSV1() (format *formatFSV1) {
        return &formatFSV1{
                formatMetaV1: formatMetaV1{
                        Version: formatMetaVersionV1,
                        Format:  formatBackendFS,
                },
                FS: struct {
                        Version string
                }{
                        Version: formatFSVersionV1,
                },
        }
}
```

Current form looks simpler:
```
func newFormatFSV1() (format *formatFSV1) {
        f := &formatFSV1{}
        f.Version = formatMetaVersionV1
        f.Format = formatBackendFS
        f.FS.Version = formatFSVersionV1
        return f
}
```

Reply:
General rule is struct literal but this style is also okay, take a look at the pkg/http for such usage.

Reply:
@krishnasrinivas Got why its complex here.

Reply:
I understand why the current implementation is complex, but the root of the problem seems to be the choice to use an anonymous struct in the definition:

```go
type formatFSV1 struct {
	formatMetaV1
	FS struct {
		Version string `json:"version"`
	} `json:"fs"`
}
```

Let's say the structure was named `FS`, then you could write like:

```go
func newFormatFSV1() (format *formatFSV1) {
    return &formatFSV1{
        formatMetaV1{
            Version: formatMetaVersionV1,
            Format: formatBackendFS,
        },
        FS{
            Version: formatFSVersionV1,
        },
    }
}
``` 

Anyway, I am not going to hold back the PR due to this suggestion. But I think the Truncate and seek to 0 is required.

Reply:
It was decided during design discussions on this format for the struct, we can discuss again and change it in a different PR if needed.